Add SessionID to the name of WorkQueue in NetworkStorageManager

szewai opened the pull request:
#### 0127ad61a7566021f2546b75fcba9ed8872b05f9
<pre>
Add SessionID to the name of WorkQueue in NetworkStorageManager
<a href="https://bugs.webkit.org/show_bug.cgi?id=254483">https://bugs.webkit.org/show_bug.cgi?id=254483</a>
rdar://107243681

Reviewed by Youenn Fablet.

With this, we will be able to know whether a session is ephemeral and whether it is created with identifier from crash
logs. This patch also adds error logs when two NetworkStorageManagers use the same path, as such use could lead to crash
shown in rdar://problem/104754030.

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::createNetworkStorageManager):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::activePaths):
(WebKit::NetworkStorageManager::create):
(WebKit::NetworkStorageManager::NetworkStorageManager):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:

Canonical link: <a href="https://commits.webkit.org/262181@main">https://commits.webkit.org/262181@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57437e7c4c005a2864e11a1988adf2ce420fbf97

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/815 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/839 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/869 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1115 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/723 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/887 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/922 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/944 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/823 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/773 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/773 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1050 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/764 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/772 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/737 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/787 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1766 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/774 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/745 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/722 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/772 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/193 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/787 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -141,6 +148,14 @@ NetworkStorageManager::NetworkStorageManager(PAL::SessionID sessionID, IPC::Conn
+        if (!existingManager)
+            activePaths().set(path, this);
+        else
+            RELEASE_LOG_ERROR(Storage, "%p - NetworkStorageManager::NetworkStorageManager path for session %" PRIu64 " is already in use by session %" PRIu64, this, m_sessionID.toUInt64(), existingManager->sessionID().toUInt64());
```
You can probably write it as:
```
auto addResult = activePaths().add(path, *this);
RELEASE_LOG_ERROR_IF(! addResult.isNewEntry, ...)
```

We could probably only do this work if RELEASE_LOG is defined, ditto for activePaths().

And maybe add a FIXME: remove once rdar.... is fixed

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -123,14 +129,15 @@ static void deleteEmptyOriginDirectory(const String& directory)
+NetworkStorageManager::NetworkStorageManager(PAL::SessionID sessionID, Markable<UUID> identifier, IPC::Connection::UniqueID connection, const String& path, const String& customLocalStoragePath, const String& customIDBStoragePath, const String& customCacheStoragePath, uint64_t defaultOriginQuota, uint64_t defaultThirdPartyOriginQuota, std::optional<double> originQuotaRatio, UnifiedOriginStorageLev ...
     : m_sessionID(sessionID)
-    , m_queue(SuspendableWorkQueue::create("com.apple.WebKit.Storage", SuspendableWorkQueue::QOS::Default, SuspendableWorkQueue::ShouldLog::Yes))
+    , m_queueName(makeString("com.apple.WebKit.Storage.", sessionID.toUInt64(), ".", static_cast<uint64_t>(identifier->data())))
```
We do not use m_queueName at all, we could probably pass it to SuspendableWorkQueue::create directly.
Also, we are exposing the whole identifier, but the log mentions it might just need to know whether there is an identifier. Should we use !!identifier? UUID is 128 bits so it seems a bit strange to just expose 64 bits.

szewai replied:
> We do not use m_queueName at all, we could probably pass it to SuspendableWorkQueue::create directly.

m_queueName is used to ensure the name string outlives the WorkQueue. Or do you mean we should create a string and never destroys it? 

> Also, we are exposing the whole identifier, but the log mentions it might just need to know whether there is an identifier. Should we use !!identifier? UUID is 128 bits so it seems a bit strange to just expose 64 bits.

The same identifier will infer the same path, so it could be helpful if we only get crash file but not log file. I tried 128 bit but it's a bit too long; I guess I could include full 128 bit.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -141,6 +148,14 @@ NetworkStorageManager::NetworkStorageManager(PAL::SessionID sessionID, IPC::Conn
+    if (!path.isEmpty()) {
+        auto existingManager = activePaths().get(path);
+        if (!existingManager)
+            activePaths().set(path, this);
```
*this?

szewai committed:
Add SessionID to the name of WorkQueue in NetworkStorageManager
https://bugs.webkit.org/show_bug.cgi?id=254483
rdar://107243681

Reviewed by Youenn Fablet.

With this, we will be able to know whether a session is ephemeral and whether it is created with identifier from crash
logs. This patch also adds error logs when two NetworkStorageManagers use the same path, as such use could lead to crash
shown in rdar://problem/104754030.

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::createNetworkStorageManager):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::activePaths):
(WebKit::NetworkStorageManager::create):
(WebKit::NetworkStorageManager::NetworkStorageManager):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:

Canonical link: https://commits.webkit.org/262181@main
